midend/flattenLogMsg.cpp

```diff
@@ -0,0 +1,198 @@
+                    nm += curResult.second;
+                    result.first.insert(result.first.end(), curResult.first.begin(),
+                                        curResult.first.end());
+                } else {
+                    result.first.push_back(new IR::NamedExpression(expr->srcInfo, newName(),
+                                                                   newMember));
+                    nm += std::string("{}");
+                }
+                if (result.second.length() > 1)
+                    result.second += ",";
+                result.second += nm;
+            }
+            result.second += std::string(")");
+            return result;
+        }
+        result.first.push_back(new IR::NamedExpression(expr->srcInfo, newName(), expr));
+        result.second = curName + std::string(" : {}");
+        return result;
+    }
+    auto* structExpr = expr->to<IR::StructExpression>();
+    std::string strResult;
+    for (auto namedExpr : structExpr->components) {
+        size_t n = strParam.find("{}");
+        if (n != std::string::npos) {
+            strResult += strParam.substr(0, n);
+            strParam = strParam.substr(n + 2);
+        }
+        exprType = typeMap->getType(namedExpr->expression, true);
+        if (exprType->is<IR::Type_StructLike>()) {
+            std::string nm;
+            if (curName.length()) {
+                nm = curName + std::string(".");
+            }
+            nm += namedExpr->name.name.c_str();
+            auto innerFields = unfoldStruct(namedExpr->expression, "", nm);
+            result.first.insert(result.first.end(), innerFields.first.begin(),
+                                innerFields.first.end());
+            strResult += innerFields.second;
+        } else {
+            strResult += "{}";
+            result.first.push_back(
+                new IR::NamedExpression(namedExpr->srcInfo, newName(), namedExpr->expression));
+        }
+    }
+    result.second = strResult + strParam;
+    return result;
+}
+
+bool FindTypeInLogMsgForReplace::preorder(const IR::MethodCallStatement* methodCallStatement) {
+    if (hasStructInParameter(methodCallStatement)) {
+        auto* newMethodCall = prepareLogMsgStatement(methodCallStatement);
+        createReplacement(newMethodCall->methodCall->typeArguments->at(0)->
+                          to<IR::Type_Struct>());
+    }
+    return false;
+}
+
+void FindTypeInLogMsgForReplace::createReplacement(const IR::Type_StructLike* type) {
+    if (replacement.count(type->name)) {
+        return;
+    }
+    replacement.emplace(type->name, type);
+}
+
+/////////////////////////////////
+
+const IR::Node* ReplaceLogMsg::preorder(IR::P4Program* program) {
+    if (findTypeInLogMsgForReplace->empty()) {
+        // nothing to do
+        prune();
+    }
+    return program;
+}
+
+const IR::Node* ReplaceLogMsg::postorder(IR::Type_Struct* typeStruct) {
```

Comment:
why do you replace the type? If this type is used in other places besides the logmsg call it should remain in the program. So you should append the new type, not replace the old one.

Reply:
Could you, please, suggest a place where is I can to insert new type?

Reply:
I think that this type couldn't be used in other places, because it should be unique for each log_msg function and each log_msg function will be replace for new one.